src/processor/fingerprint.cpp

```diff
@@ -0,0 +1,283 @@
+
+    ddwaf_object res;
+    auto [ptr, size] = buffer.move();
+    ddwaf_object_stringl_nc(&res, ptr, size);
+
+    return res;
+}
+
+// Retur true if the first argument is less than (i.e. is ordered before) the second
+bool str_casei_cmp(std::string_view left, std::string_view right)
+{
+    auto n = std::min(left.size(), right.size());
+    for (std::size_t i = 0; i < n; ++i) {
+        auto lc = ddwaf::tolower(left[i]);
+        auto rc = ddwaf::tolower(right[i]);
+        if (lc != rc) {
+            return lc < rc;
+        }
+    }
+    return left.size() <= right.size();
+}
+
+void normalize_string(std::string_view key, std::string &buffer, bool trailing_separator)
+{
+    // Clear should not deallocate...
+    // TODO: verify
+    buffer.clear();
+
+    if (buffer.capacity() < key.size()) {
+        // Add space for the extra comma, just in case
+        buffer.reserve(key.size() + 1);
+    }
+
+    for (auto c : key) {
+        if (c == ',') {
+            buffer.append(R"(\,)");
+        } else {
+            buffer.append(1, ddwaf::tolower(c));
+        }
+    }
+
+    if (trailing_separator) {
+        buffer.append(1, ',');
+    }
+}
+
+void string_hash_field::operator()(string_buffer &output)
+{
+    if (value.empty()) {
+        return;
+    }
+
+    cow_string value_lc{value};
+    transformer::lowercase::transform(value_lc);
+
+    sha256_hash hasher;
+    hasher << static_cast<std::string_view>(value_lc);
+
+    hasher.write_digest(output.subspan<8>());
+}
+
+void key_hash_field::operator()(string_buffer &output)
+{
+    if (value.type != DDWAF_OBJ_MAP or value.nbEntries == 0) {
+        return;
+    }
+
+    std::vector<std::string_view> keys;
+    keys.reserve(value.nbEntries);
+
+    for (unsigned i = 0; i < value.nbEntries; ++i) {
+        const auto &child = value.array[i];
+
+        std::string_view key{
+            child.parameterName, static_cast<std::size_t>(child.parameterNameLength)};
+
+        keys.emplace_back(key);
+    }
+
+    std::sort(keys.begin(), keys.end(), str_casei_cmp);
+
+    sha256_hash hasher;
+    std::string normalized;
+    for (unsigned i = 0; i < keys.size(); ++i) {
+        normalize_string(keys[i], normalized, (i + 1) < keys.size());
+        hasher << normalized;
+    }
+
+    hasher.write_digest(output.subspan<8>());
+}
+
+} // namespace
+
+// NOLINTNEXTLINE(readability-convert-member-functions-to-static)
+std::pair<ddwaf_object, object_store::attribute> http_endpoint_fingerprint::eval_impl(
+    const unary_argument<std::string_view> &method, const unary_argument<std::string_view> &uri_raw,
+    const unary_argument<const ddwaf_object *> &query,
+    const unary_argument<const ddwaf_object *> &body, ddwaf::timer &deadline) const
+{
+    if (deadline.expired()) {
+        throw ddwaf::timeout_exception();
+    }
+
+    // Strip query parameter from raw URI
+    auto stripped_uri = uri_raw.value;
+    auto query_idx = stripped_uri.find_first_of('?');
```

Comment:
That doesn't feel terribly robust

Reply:
Should be robust enough as the `?` character isn't allowed in the [URI path](https://datatracker.ietf.org/doc/html/rfc3986#section-3.1:~:text=zero%2Dlength%20segment%20without%20any%20colon%20%22%3A%22-,pchar,-%3D%20unreserved%20/%20pct%2Dencoded%20/%20sub%2Ddelims%20/%20%22%3A%22%20/%20%22%40%22%0A%0A%20%20%20query).

Reply:
Ah indeed, forgot about fragments entirely.